speculos/mcu/apdu.py

```diff
@@ -8,39 +10,48 @@
 import errno
 import logging
 import socket
+from typing import Optional
 
+from .display import IODevice, DisplayNotifier
 
-class ApduServer:
-    def __init__(self, host='127.0.0.1', port=9999, hid=False):
-        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        self.s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        self.s.bind((host, port))  # lgtm [py/bind-socket-all-network-interfaces]
-        self.s.listen()
 
-        self.client = None
+class ApduServer(IODevice):
+    def __init__(self, host: str = '127.0.0.1', port: int = 9999, hid: bool = False):
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.socket.bind((host, port))  # lgtm [py/bind-socket-all-network-interfaces]
+        self.socket.listen()
+        self.client: Optional[ApduClient] = None
 
-    def can_read(self, s, screen):
-        assert self.s.fileno() == s
+    @property
+    def file(self):
+        return self.socket
 
-        c, _ = self.s.accept()
+    def can_read(self, s: int, screen: DisplayNotifier):
+        assert self.fileno == s
+        c, _ = self.file.accept()
         self.client = ApduClient(c)
         screen.add_notifier(self.client)
 
-    def forward_to_client(self, packet):
+    def forward_to_client(self, packet: bytes):
         if self.client is not None:
             self.client.forward_to_client(packet)
```

Comment:
I've seen this type of code multiple times, is it possible to move it to IODevice ?

Reply:
On a second read probably not

Reply:
Maybe on the `can_read` method (I feel some socket devices share some equivalent code), but I think this refacto would a bit too complicated for now. It won't be just moving code to some other place, it would need to actually understand it.